pkg/core/writer/splunk/splunk.go

```diff
@@ -0,0 +1,240 @@
+package splunk
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/signalfx/golib/v3/datapoint"
+	"github.com/signalfx/golib/v3/event"
+	"github.com/signalfx/signalfx-agent/pkg/core/common/httpclient"
+	log "github.com/sirupsen/logrus"
+)
+
+// Writer posts data to Splunk HTTP Event Collector.
+type Writer struct {
+	httpClient    *http.Client
+	url           string
+	hostname      string
+	token         string
+	source        string
+	sourceType    string
+	index         string
+	skipTLSVerify bool
+	events        chan interface{}
+	sendQueue     chan []interface{}
+}
+
+// Build a Splunk Writer.
+func Build(url string, token string, source string, sourceType string, index string, skipTLSVerify bool, hostname string) (Writer, error) {
+	handler := Writer{
+		url:           url,
+		token:         token,
+		source:        source,
+		sourceType:    sourceType,
+		index:         index,
+		skipTLSVerify: skipTLSVerify,
+		hostname:      hostname,
+	}
+	err := handler.init()
+	return handler, err
+}
+
+func (h *Writer) init() error {
+	httpConfig := httpclient.HTTPConfig{
+		SkipVerify: h.skipTLSVerify,
+		UseHTTPS:   strings.HasPrefix(h.url, "https"),
+	}
+
+	httpClient, err := httpConfig.Build()
+	if err != nil {
+		return err
+	}
+	h.httpClient = httpClient
+
+	if h.hostname == "" {
+		h.hostname, _ = os.Hostname()
+	}
+	h.sendQueue = make(chan []interface{})
+	h.events = make(chan interface{})
+	go func() {
+		eventsToSend := make([]interface{}, 0)
+		ticker := time.NewTicker(1 * time.Second)
+		for {
+			select {
+			case e := <-h.events:
+				eventsToSend = append(eventsToSend, e)
+				if len(eventsToSend) > 50 {
+					oldEvents := eventsToSend
+					eventsToSend = make([]interface{}, 0)
+					h.sendQueue <- oldEvents
+				}
+				break
+			case <-ticker.C:
+				if len(eventsToSend) > 0 {
+					oldEvents := eventsToSend
+					eventsToSend = make([]interface{}, 0)
+					h.sendQueue <- oldEvents
+				}
+			}
+		}
+	}()
+	go func() {
+		for {
+			events := <-h.sendQueue
```

Comment:
This sending model is pretty similar to the original model we used for the agent (buffer a bunch of data, and send them out after a certain batch size is hit or after a certain period of time has passed).  We have since switched to something a bit more sophisticated since then due to limitations that we had with that at high volumes of data generation.

This for loop here can only handle sending one batch at a time since logEvents is synchronous.  That means you are primarily limited by the RTT of the request to Splunk, which is uncontrollable by the agent. Also since you are limiting the batch size to 50 that means you have a max event rate of `1/<RTT in seconds> * 50` events per second.

We can discuss what the other writer in the agent does and whether it is appropriate for this when we talk later today.

Reply:
Definitely open to anything - and it looks like you have a better model for a writer in hand, so let's use that.